src/Illuminate/Contracts/Database/Eloquent/Castable.php

```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace Illuminate\Contracts\Database\Eloquent;
+
+interface Castable
+{
+    /**
+     * Get a given attribute from the model.
+     *
+     * @param  mixed  $value
+     * @return mixed
+     */
+    public function get($value = null);
```

Comment:
Is it a good idea to also pass the model instance as a parameter? Sometimes when writing an accessor I need to check on another field from the model to determine the correct value. I guess this might be too specific and would be better to write custom accessors for cases like this.

Reply:
Or maybe inject the current model in the cast constructor. (In this case disregard my other comment about keeping the cast instances on a static array and resolving the cast from the container)

Reply:
If I understood you correctly, then I already came across this case when writing tests, when I needed to display the state of a specific field.
I agree, now we can only track the result, but the result of which is incomprehensible.

Now I’ll think about how to do it right and add the necessary code.

Reply:
I did.

Added four methods: two for receiving and two for setting keys and values (`setKeyName`, `getKeyName`, `setOriginalValue`, `getOriginalValue`)

Reply:
See [commit](https://github.com/laravel/framework/pull/30958/commits/ce3985ac7c478bc5f2699a66e2ef1e4feeef06d4)

Reply:
I was thinking how to improve my previous [commit](https://github.com/laravel/framework/pull/30958/commits/ce3985ac7c478bc5f2699a66e2ef1e4feeef06d4).

### 1
Interface:
```php
namespace Illuminate\Contracts\Database\Eloquent;

interface Castable
{
    public function get($key, $value = null);

    public function set($key, $value = null);
}
```

`hasCasts` trait:
```php
    protected function fromCustomCastable($key, $value = null)
    {
        return $this
            ->normalizeHandlerToCallable($key)
            ->get($key, $value);
    }

    protected function toCustomCastable($key, $value = null)
    {
        return $this
            ->normalizeHandlerToCallable($key)
            ->get($key, $value);
    }
```

### 2
Interface:
```php
namespace Illuminate\Contracts\Database\Eloquent;

interface Castable
{
    public static function get($key, $value = null);

    public static function set($key, $value = null);
}
```

`hasCasts` trait:
```php
    protected function fromCustomCastable($key, $value = null)
    {
        return call_user_func(
            [$this->getCast($key), 'get'],
            $key, $value
        );
    }

    protected function toCustomCastable($key, $value = null)
    {
        return call_user_func(
            [$this->getCast($key), 'set'],
            $key, $value
        );
    }
```

What do you think?

Reply:
Hmm... I don't think I preferred the previous API, sorry If my comment wasn't clear enough, might be the language barrier as I am not a native English speaker. 

Will try to speak through code. I meant something like this:

Sometimes you need to use another column to calculate a field's value

~~~php 
class Product extends Model {
  public function getPriceAttribute($value) {
    return $value * $this->promotional_discount;
  }
}
~~~

So if the custom cast receives the model somehow one could refactor to this:

~~~php
class Product extends Model {
  public $casts = ['price' => ProductPriceCast::class],
}
~~~

If we inject the current instance through the constructor we could write the cast as such:

~~~php
class ProductPriceCast implements Cast {
  public function __construct(Model $model) {
    $this->model = $model;
  }

  public function get($value = null) {
    return $value * $model->promotional_discount;
  }

  public function set($value = null) {
    return $value;
  }
}
~~~

But maybe this is a overkill, and we could use a custom accessor for this.

On the other hand, after reading through the previous PR comments, I guess this could allow to have Value Objects based on multiple columns:

~~~php
class Contact extends Model {
  public $casts = ['address' => AddressCast::class],
}
~~~

~~~php
class AddressCast implements Cast {
  public function __construct(Model $model) {
    $this->model = $model;
  }

  public function get($value = null) {
    return new AddressValueObject($this->model->street, $this->model->zip, $this->model->city, $this->model->state);
  }

  public function set($address = null) {
    $this->model->street = $address->getStreet();
    $this->model->zip = $address->getZip();
    $this->model->city = $address->getCity();
    $this->model->state = $address->getState();
  }
}
~~~

Reply:
I guess I might have clicked on the wrong button, I didn't mean to start a review